src/realm/object-store/dictionary.cpp

```diff
@@ -41,6 +41,13 @@ Dictionary::Dictionary(std::shared_ptr<Realm> r, const realm::Dictionary& dict)
     REALM_ASSERT(m_dict);
 }
 
+Dictionary::Dictionary(std::shared_ptr<Realm> r, CollectionBasePtr coll)
+    : Collection(std::move(r), std::move(coll))
+    , m_dict(dynamic_cast<realm::Dictionary*>(m_coll_base.get()))
+{
+    REALM_ASSERT(m_dict);
+}
```

Comment:
I guess I disagree a bit about this design. It's unfortunate that we could get into an invalid internal state of `Dictionary` in release builds.

Would it make sense to do the cast in the implementation of `freeze()` after calling `import_copy_of()`, and potentially throw a `LogicError` if it fails for some reason?